smart-contract-extensions/test/hooks/discountCodeHook.js

```diff
@@ -0,0 +1,228 @@
+          from: keyBuyer,
+        }
+      ),
+      'INSUFFICIENT_VALUE'
+    )
+  })
+
+  it('you can buy at a discount with the code', async () => {
+    await lock.purchase(
+      '0',
+      keyBuyer,
+      constants.ZERO_ADDRESS,
+      await getDataField(lock, keyBuyer, 'joerogan'),
+      {
+        // This code cuts the price in half
+        value: keyPrice.div(2).toFixed(),
+        from: keyBuyer,
+      }
+    )
+  })
+
+  it('the available discount is not more than expected', async () => {
+    await reverts(
+      lock.purchase(
+        '0',
+        keyBuyer,
+        constants.ZERO_ADDRESS,
+        await getDataField(lock, keyBuyer, 'joerogan'),
+        {
+          value: keyPrice
+            .div(2)
+            .minus(1)
+            .toFixed(),
+          from: keyBuyer,
+        }
+      ),
+      'INSUFFICIENT_VALUE'
+    )
+  })
+
+  it('you can still leave a tip when purchasing with a code', async () => {
+    await lock.purchase(
+      '0',
+      keyBuyer,
+      constants.ZERO_ADDRESS,
+      await getDataField(lock, keyBuyer, 'joerogan'),
+      {
+        value: keyPrice
+          .div(2)
+          .plus(web3.utils.toWei('0.1', 'ether'))
+          .toFixed(),
+        from: keyBuyer,
+      }
+    )
+  })
+
+  it('different codes may offer different discounts', async () => {
+    const purchasePrice = await lock.purchasePriceFor(
+      keyBuyer,
+      constants.ZERO_ADDRESS,
+      await getDataField(lock, keyBuyer, 'unlock 25'),
+      {
+        from: keyBuyer,
+      }
+    )
+
+    // This code reduces the price by 25%
+    assert.equal(purchasePrice.toString(), keyPrice.times(0.75).toFixed())
+  })
+
+  it('non-admins cannot modify codes', async () => {
+    await reverts(
+      hookContract.addCodes([getCodeAddress(lock, 'fail')], [5000], {
+        from: keyBuyer,
+      }),
+      'WhitelistAdminRole: caller does not have the WhitelistAdmin role'
+    )
+  })
+
+  describe('codes can be removed', () => {
+    beforeEach(async () => {
+      await hookContract.addCodes([getCodeAddress(lock, 'joerogan')], [0], {
+        from: lockCreator,
+      })
+    })
+
+    it('now the code no longer offers any discount', async () => {
+      const purchasePrice = await lock.purchasePriceFor(
+        keyBuyer,
+        constants.ZERO_ADDRESS,
+        await getDataField(lock, keyBuyer, 'joerogan'),
+        {
+          from: keyBuyer,
+        }
+      )
+
+      assert.equal(purchasePrice.toString(), keyPrice.toFixed())
+    })
+  })
+
+  describe('the hook contract may be removed', () => {
+    beforeEach(async () => {
+      await lock.setEventHooks(constants.ZERO_ADDRESS, constants.ZERO_ADDRESS, {
+        from: lockCreator,
+      })
+    })
+
+    it('now the code no longer offers any discount', async () => {
+      const purchasePrice = await lock.purchasePriceFor(
+        keyBuyer,
+        constants.ZERO_ADDRESS,
+        await getDataField(lock, keyBuyer, 'joerogan'),
```

Comment:
Should this mzybe be tested with the `unlock 25` code, as the `joerogan` code was already set to 0?